### ISSUE TITLE
Upgrade minSDK to 21 until end of 2020

### ISSUE THREAD

**Issue by moving-bits** [cgeo/cgeo#8120]
(Edit: We support down to minSDK 16, but the message is basically the same.)

Currently we support down to minSDK 19 (which is Android 4.4). This blocks at least OK.HTTP from upgrading to the v4 branch. v3 branch is deprecated and will get crucial security fixes only, and only until end of 2020 (see https://square.github.io/okhttp/#requirements). (There are probably other dependencies blocked by that, haven't checked, though.)

As of May 2019 the market share of Android 4.4 was down to 6.9%, it should be less now, and even less until end of 2020.
According to the stats Lineflyer sent around in November 2019, it's even less among the c:geo users, a mere 2.3% using Android 4.4 with c:geo.

So I propose to deprecate Android 4.4 for c:geo during the next months, and revoke Android 4.4 support until fall of 2020.

**Comment by bekuno** [comment]
If necessary, we can define a version as "legacy" and only fix website changes on there.

**Comment by bekuno** [comment]
see https://github.com/cgeo/cgeo/issues/8135#issuecomment-589998434

**Comment by Lineflyer** [comment]
AFAIR our minSDK is 16, not 19.

**Comment by moving-bits** [comment]
> AFAIR our minSDK is 16, not 19.

You're right - we currently support minSDK 16, which is Android 4.1

Android versions up to 4.0 have a share of 0.6%, according to Googles distribution dashboard as of May 19. Android versions up to 4.4.x had a share of 10.7% at that time, should be less now.

**Comment by jonas-koeritz** [comment]
Going above 18 would enable us to use the KeyStorw to securely encrypt user credentials in storage as well.

**Comment by Lineflyer** [comment]
Some statistics taken from Google Play as of today:


Android Version | Active installs
-- | --
Android   4.1 | 2637
Android 4.2 | 3076
Android 4.3 | 1585
Android 4.4 | 15287

Total active installations: 800k

So the percentage of Android 4.x is about 2,8%.
BTW: Active installs does not provide any information about usage, just that c:geo is confirmed to be installed on those devices within the last 7 (or 14) days.

**Comment by mirabilos** [comment]
How about asking the OKHTTP people to support 4.4 then?

**Comment by jonas-koeritz** [comment]
@mirabilos I think they couldn't care less as we're not using the latest version of the library. I don't know of their release cycles and stuff but I assume backporting fixes for ancient Android versions is a lot to ask for.

Maybe we could "tweak" the settings we're using with OKHTTP to stop trying SSL3? I think there is an option to disable specific protocols.

**Comment by Lineflyer** [comment]
> How about asking the OKHTTP people to support 4.4 then?
This probably refers to issue #7497 rather than this issue (while it is connected somehow). Please continue the discussion over there. The outcome however might then influence the further proceeding regarding this issue.

I would personally vote for:
- Finishing our ongoing release at first
- Copying release to legacy
- Try to apply a workaround for #7497 on legacy
- Publish a legacy version

**Comment by mirabilos** [comment]
Jonas Köritz dixit:

>Maybe we could "tweak" the settings we're using with OKHTTP to stop
>trying SSL3? I think there is an option to disable specific protocols.

That would be best.


Lars dixit:

>- Publish a legacy version

Hm, I’m using nightlies though (mostly due to the often GC changes).

**Comment by moving-bits** [comment]
Stumbled upon by an newsletter article that Android Studio has an overview over current (?) Android version usage statistics. As of today (April 13th) it is as follows:

![image](https://user-images.githubusercontent.com/3754370/79100419-bb41a880-7d66-11ea-99f8-0843832f26c7.png)

According to this switching to Android 5 (API level 21) as minimum requirement would exclude about 5.9% as seen by Play store overall (and even less according to Lineflyer's numbers regarding c:geo).

**Comment by mirabilos** [comment]
moving-bits dixit:

>According to this switching to Android 5 (API level 21) as minimum
>requirement would exclude about 5.9% as seen by Play store overall (and
					 ^^^^^^^^^^^^^^^^^^

I’d expect many users of older versions to be on custom ROMS without
Google stuff installed (it is simply too slow), so the numbers are
much higher.

As someone personally affected, I would be decidedly unhappy.

**Comment by moving-bits** [comment]
@mirabilos 
I do understand that personally, but the day will come where we have to drop older Android versions. :-(

The extra mileage to keep things in a consistent way across such a broad variety of Android versions is pretty huge, especially when trying to do things which don't have a standardized way of doing (like seen again lately while working on #8180). Android platform and API have changed quite a bit during the nearly eight years that lay between Android 4.1 and Android 10. Also the testing effort is tremendous in those cases, as we have to cross check along 13 versions (if we just count point releases).

An additional reason is that we depend on third party libraries in some (but relevant) places. Many of them do not support older Android versions anymore, so we need to stay at some more or less unmaintained legacy version, opening c:geo for relevant security bugs and also cut off from further development.

Not to speak of new features appearing in newer versions of Android, some of which may be useful to c:geo.

To cut a long story short: The day will come where we have to drop at least Android 4.x, and it's not unlikely that 2020 will be the year, as a relevant library we use drops it already scarce support they still have for their legacy version we are currently forced to use.

**Comment by Lineflyer** [comment]
> As someone personally affected, I would be decidedly unhappy.

As mentioned somewhere above, we could cut off a legacy version for 4.x devices which can run without new features and with important bugfixes (as long as we decide to drop it somewhen significantly later).

**Comment by bekuno** [comment]
My suggestion is to stay on our plan to drop support for API < 21 to the end of the year (or in the 4th quarter). We should raise the info for the affected user ~3 month before with a planned timeline. 
In the meantime we can prepare some PR for it and track them in a project.

**Comment by Lineflyer** [comment]
> My suggestion is to stay on our plan to drop support for API < 21 to the end of the year (or in the 4th quarter). We should raise the info for the affected user ~3 month before with a planned timeline.

Yes, but nothinh speaks against cutting out the lower SDKs to a legacy version already now. 
This way the development on the "normal" branches can start right away.

**Comment by okainov** [comment]
Probably duplicate of #8248 which was fixed already, so should it be closed as well?

**Comment by mirabilos** [comment]
@Lineflyer is there anywhere I can see the progress of the “legacy” version?

I haven’t updated my c:geo in six months for fear of it stopping to work, plus, OC.DE is still not working there (the legacy version could revert to a working OKHTTP release)…

**Comment by moving-bits** [comment]
We plan to cut off a new legacy version from the release we are currently preparing, and then apply the #7497 fix to it.

**Comment by moving-bits** [comment]
Min SDK ist set to 21 on master branch, so let's close here.
Legacy version should get a separate issue for further tracking.
